Add functional test for IPC interface

sipa opened the pull request:
This adds support to the functional test framework to run the multiprocess `bitcoin-node` binary, and then tests it in a new `interface_ipc.py` functional test through the `pycapnp` module.

ryanofsky committed:
test: Provide path to `bitcoin` binary

Set new `BitcoinTestFramework.binary_paths.bitcoin_bin` property with path to
the `bitcoin` wrapper binary. This allows new tests for `bitcoin-mine` in
#30437 and `bitcoin-cli` in #32297 to find the `bitcoin` binary and call
`bitcoin -m` to start nodes with IPC support. This way the new tests can run
whenever the ENABLE_IPC build option is enabled, instead of only running when
the `BITCOIN_CMD` environment variable is set to `bitcoin -m`

ryanofsky committed:
test: add is_ipc_compiled() and skip_if_no_ipc() functions

Expose ENABLE_IPC build option to functional tests so new tests can test
IPC-only features.

ryanofsky commented:
Concept ACK. 27bd67857dd0e145d7539b949df2b7aaa53c6b1b is really nice and simpler than I would have expected.

In case it's useful #30437 adds an [interface_ipc_mining.py](https://github.com/ryanofsky/bitcoin/blob/pr/mine/test/functional/interface_ipc_mining.py) test, and #32297 adds an [interface_ipc_cli.py](https://github.com/ryanofsky/bitcoin/blob/pr/ipc-cli/test/functional/interface_ipc_cli.py) test which are similar to this test and have similar test framework updates. Code from those PRs might be useful here or vice versa.

Sjors commented on test/functional/interface_ipc.py:
```diff
@@ -0,0 +1,86 @@
+            opts.useMempool = True
+            opts.blockReservedWeight = 4000
+            opts.coinbaseOutputMaxAdditionalSigops = 0
+            template = mining.result.createNewBlock(opts)
```
It would be great to cover the `waitNext()` method on `template`.

Sjors reviewed (commented):
On macOS 15.6 with capnp 1.2.0 installed via Homebrew and Python 3.10.14 installed via PyEnv I get:

```
pip3 install pycapnp
...
Successfully installed pycapnp-2.0.0

cmake -B build
cmake --build build
...

% build/test/functional/interface_ipc.py
2025-08-17T07:05:49.321522Z TestFramework (INFO): PRNG seed is: 6039316184654231320
2025-08-17T07:05:49.322566Z TestFramework (INFO): Initializing test directory /var/folders/8p/qqdl2fsj7rd6q096jtmsfydc0000gn/T/bitcoin_func_test_z3jtr5u1
2025-08-17T07:05:50.595292Z TestFramework (ERROR): Unexpected exception
Traceback (most recent call last):
  File "/Users/sjors/dev/bitcoin/test/functional/test_framework/test_framework.py", line 200, in main
    self.run_test()
  File "/Users/sjors/dev/bitcoin/build/test/functional/interface_ipc.py", line 82, in run_test
    self.run_echo_test()
  File "/Users/sjors/dev/bitcoin/build/test/functional/interface_ipc.py", line 63, in run_echo_test
    asyncio.run(capnp.run(async_routine()))
  File "/Users/sjors/.pyenv/versions/3.10.14/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/Users/sjors/.pyenv/versions/3.10.14/lib/python3.10/asyncio/base_events.py", line 649, in run_until_complete
    return future.result()
  File "capnp/lib/capnp.pyx", line 1944, in run
  File "capnp/lib/capnp.pyx", line 1945, in capnp.lib.capnp.run
  File "/Users/sjors/dev/bitcoin/build/test/functional/interface_ipc.py", line 57, in async_routine
    init, ctx = await self.make_capnp_init_ctx()
  File "/Users/sjors/dev/bitcoin/build/test/functional/interface_ipc.py", line 45, in make_capnp_init_ctx
    modules = self.capnp_modules()
  File "/Users/sjors/dev/bitcoin/build/test/functional/interface_ipc.py", line 38, in capnp_modules
    "proxy": capnp.load(str(mp_dir / "mp" / "proxy.capnp"), imports=imports),
  File "capnp/lib/capnp.pyx", line 4385, in capnp.lib.capnp.load
  File "capnp/lib/capnp.pyx", line 3574, in capnp.lib.capnp.SchemaParser.load
capnp.lib.capnp.KjException: mp/proxy.capnp:6: failed: Import failed: /capnp/c++.capnp
```

Maybe related: https://github.com/capnproto/pycapnp/issues/377 (though `pycapnp-1.3.0` doesn't work at all, since it doesn't have the `run` method)

Suggested test doc improvement:

```diff
diff --git a/test/README.md b/test/README.md
index 37f2c07217..3b2979cdf6 100644
--- a/test/README.md
+++ b/test/README.md
@@ -34,6 +34,9 @@ The ZMQ functional test requires a python ZMQ library. To install it:
 - on Unix, run `sudo apt-get install python3-zmq`
 - on mac OS, run `pip3 install pyzmq`

+The IPC functional test requires a python IPC library. To install it:
+
+- `pip3 install pycapnp`

 On Windows the `PYTHONUTF8` environment variable must be set to 1:
```

janb84 commented:
> @janb84 Yeah, the include paths used in the python test are a guess. It'd be nice to pass through the location of the capnp installation from the build system to the tests, but I don't know how.

I'm also looking into it to get it to work, just a heads-up that there is (possibly) some trouble on nixos/ nix-shells

ryanofsky committed:
test: Add TestNode ipcbind option

With this change, tests can specify `self.extra_init = [{ipcbind: True}]` to
start a node listening on an IPC socket, instead of needing to choose which
node binary to invoke and what `self.extra_args=[["-ipcbind=..."]]` value to
pass to it.

The eliminates boilerplate code #30437 (interface_ipc_mining.py), #32297
(interface_ipc_cli.py), and #33201 (interface_ipc.py) previously needed in
their test setup.

ismaelsadeeq reviewed (commented):
Approach ACK 

First pass initial comments

josibake reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/33201/commits/3f7ff4849e6165a6bcc4d5977c35a467fd7fc232

Thanks for picking this up! I think the docs and multiplatform support still need a bit of work but that can be tackled in a follow-up(s), considering the main benefit I see here is getting a basic functional test in place that can be extended later.

ismaelsadeeq reviewed (approved):
ACK 3f7ff4849e6165a6bcc4d5977c35a467fd7fc232

Sjors commented on test/functional/interface_ipc.py:
```diff
@@ -0,0 +1,150 @@
+        cpp_capnp_dir = Path(capnp.__path__[0]).parent
+        src_dir = Path(self.config['environment']['SRCDIR']) / "src"
+        mp_dir = src_dir / "ipc" / "libmultiprocess" / "include"
+        imports = ["/usr/include", str(cpp_capnp_dir), str(src_dir), str(mp_dir)]
```
As of cd86ddaa1dac104415ab16cc8068fd50d068c12f the test passes for me on macOS with Python 3.10.14 with pycapnp installed via pip and capnp installed via homebrew.

I also tested Python 3.12.11. With Python 3.13.17 `pip3 install pycapnp` fails. We can worry about that later.

Sjors commented:
If this PR is to be based on #31802 or #33190 then it should be marked draft. Otherwise it could be made independent by dropping 7a7f11933b83d290129377523cc4a3cfc45fe9ef and instead enabling IPC for one or more CI machines that run functional tests. It should be straightforward to rebase #31802 as long as you leave `ci/test/00_setup_env_i686_multiprocess.sh` alone (due to 789f7195b2acbc3e50ee75b2b5a88af058cf7b14). My preference would be to base it on #31802 because that gives the widest test coverage, including depends builds.

Sjors commented on test/functional/test_framework/test_framework.py:
```diff
@@ -964,6 +964,13 @@ def skip_if_no_py_sqlite3(self):
+    def skip_if_no_py_capnp(self):
+        """Attempt to import the capnp package and skip the test if the import fails."""
+        try:
+            import capnp  # type: ignore[import] # noqa: F401
```
Ah yes, now it's failing with a very cryptic "Tests successful"

ryanofsky commented on test/functional/test_framework/test_framework.py:
```diff
@@ -207,6 +207,12 @@ def main(self):
         except BaseException:
             self.log.exception("Unexpected exception")
             self.success = TestStatus.FAILED
+        except SystemExit as e:
+            self.log.exception(f"System exit: {e.message}")
+            self.success = TestStatus.FAILED
+        except GeneratorExit as e:
+            self.log.exception(f"Generator exit: {e.message}")
+            self.success = TestStatus.FAILED
```
IIUC it's failing because stderr is nonempty:

https://github.com/bitcoin/bitcoin/blob/22e689587a7ae02f82ad2464017731f3b23b5363/test/functional/test_runner.py#L760

because the [log](https://github.com/bitcoin/bitcoin/actions/runs/1753347607/job/48345965497) ([raw](https://productionresultssa13.blob.core.windows.net/actions-results/bfb1b205-2f78-46f1-95d9-f32ca7112d38/workflow-job-run-45b4c18f-0ff3-5fd3-8716-ea6a91129b24/logs/job/job-logs.txt?rsct=text%2Fplain&se=2025-08-19T14%3A22%3A57Z&sig=8J3CHfDQKjltz%2BnlxtXd8bm2gQPw0%2FIx8XJkLH2LMwY%3D&ske=2025-08-20T02%3A07%3A23Z&skoid=ca7593d4-ee42-46cd-af88-8b886a2f84eb&sks=b&skt=2025-08-19T14%3A07%3A23Z&sktid=398a6654-997b-47e9-b12b-9515b896b4de&skv=2025-05-05&sp=r&spr=https&sr=b&st=2025-08-19T14%3A12%3A52Z&sv=2025-05-05)) includes "[node 0] Cleaning up ipc directory '/tmp/test-ipc-xj2xyof4'" as maflcko mentioned.

Maybe this line should be changed to use stdout instead of stderr?

https://github.com/bitcoin/bitcoin/blob/530ca0bc57477b77f60af142695d9bf200ef6e3c/test/functional/test_framework/test_node.py#L232

(Sorry if this is the bug)

janb84 replied:
> 
> Maybe this line should be changed to use stdout instead of stderr?
> 
> https://github.com/bitcoin/bitcoin/blob/530ca0bc57477b77f60af142695d9bf200ef6e3c/test/functional/test_framework/test_node.py#L232
> 
> (Sorry if this is the bug)

I Could reproduce the bug locally and this change solves it for me.

<details>

Without change :

```shell
Remaining jobs: [interface_ipc.py]
1/1 - interface_ipc.py failed, Duration: 4 s

stdout:
2025-08-19T14:35:47.614991Z TestFramework (INFO): PRNG seed is: 1494337157176280909
2025-08-19T14:35:47.615519Z TestFramework (INFO): Initializing test directory /private/tmp/nix-shell-71975-4016936104/test_runner_₿_🏃_20250819_163547/interface_ipc_0
2025-08-19T14:35:49.166272Z TestFramework (INFO): Running echo test
2025-08-19T14:35:49.168265Z TestFramework (INFO): Running mining test
2025-08-19T14:35:51.247065Z TestFramework (INFO): Stopping nodes
2025-08-19T14:35:51.411476Z TestFramework (INFO): Cleaning up /private/tmp/nix-shell-71975-4016936104/test_runner_₿_🏃_20250819_163547/interface_ipc_0 on exit
2025-08-19T14:35:51.411598Z TestFramework (INFO): Tests successful


stderr:
[node 0] Cleaning up ipc directory '/private/tmp/nix-shell-71975-4016936104/test-ipc-qbv8akuk'

TEST             | STATUS    | DURATION

interface_ipc.py | ✖ Failed  | 4 s

ALL              | ✖ Failed  | 4 s (accumulated)
Runtime: 4 s
```

with change:

```shell

Temporary test directory at /private/tmp/nix-shell-71975-4016936104/test_runner_₿_🏃_20250819_163804
Remaining jobs: [interface_ipc.py]
1/1 - interface_ipc.py passed, Duration: 3 s

TEST             | STATUS    | DURATION

interface_ipc.py | ✓ Passed  | 3 s

ALL              | ✓ Passed  | 3 s (accumulated)
Runtime: 3 s

```

</details>

Sjors commented on test/functional/interface_ipc.py:
```diff
@@ -0,0 +1,166 @@
+        # as it is being called before knowing whether capnp is available).
+        self.capnp_modules = self.load_capnp_modules()
+
+    async def make_capnp_init_ctx(self):
```
088dc2c486af0ad6803d919d8114ab29d3cd652f: eventually I'd like to move this and `load_capnp_modules` into the test framework so that it's easier to expand e.g. `mining_template_verification.py` to call both `getblocktemplate proposal` and `checkBlock()`.

Sjors commented on test/functional/interface_ipc.py:
```diff
@@ -0,0 +1,166 @@
+                block3 = await self.parse_and_deserialize_block(template3, ctx)
+                assert_equal(len(block3.vtx), 2)
+                last_template = template3
+            # Wait for another, but timeout.
```
088dc2c486af0ad6803d919d8114ab29d3cd652f: nit if you have to retouch: these comments are useful as `self.log.debug()`

Sjors reviewed (approved):
ACK 088dc2c486af0ad6803d919d8114ab29d3cd652f (after base PR lands)

janb84 reviewed (commented):
ACK 088dc2c486af0ad6803d919d8114ab29d3cd652f (after base #31802)

This PR adds some functional tests for the IPC interface, setups the CI to use IPC and changes build documentation to match the changes in this PR.

Current state of the PR works on nix on macOS, builds & test run without trouble.

- build ✅
- tested ✅
- code review✅ 

small NIT if retouched, please see the remarks of DrahtBot's LLM linter.

ryanofsky reviewed (approved):
Code review ACK 088dc2c486af0ad6803d919d8114ab29d3cd652f. Looks great! Left some comments that you should feel free to ignore. I am shocked and amazed about how much more friendly the python async API is compared to the c++ API, but I guess coroutines can bring similar benefits to c++. It is also nice how this tests covers different scenarios with waitNext()

Sjors commented:
@sipa let me know if you don't have to time to rebase, in which case I'll open a fresh PR next week.

sipa committed:
tests: add functional tests for IPC interface

Co-Authored-By: ismaelsadeeq <ask4ismailsadiq@gmail.com>
Co-Authored-By: ryanofsky <ryan@ofsky.org>
Co-Authored-By: TheCharlatan <seb.kung@gmail.com>
Co-Authored-By: Sjors Provoost <sjors@sprovoost.nl>

sipa committed:
ci: enable IPC tests in CI

janb84 reviewed (commented):
re ACK 03c66ee229393335fabae1192b871f9ec68adf74

changes since last ACK:
- rebase
- addressed some NIT's

Sjors committed:
ci: test IPC on additional hosts

Install pycapnp on all (active) CI hosts which have IPC enabled and
run the functional tests.

Except for previous_releases, which uses an older version of pip
that doesn't support --break-system-packages.

Sjors commented:
ACK 235016f5b78ba9f472b56df0825690307fffc7e6

janb84 reviewed (commented):
re ACK 235016f5b78ba9f472b56df0825690307fffc7e6

changes since last ACK: 
- Rebase
- Fixes for macOS native job (pip -> pip3)
- Added IPC interface tests to more hosts.

ismaelsadeeq commented on test/functional/interface_ipc.py:
```diff
@@ -0,0 +1,169 @@
+            blockref = await mining.result.getTip(ctx)
+            assert blockref.hasResult
+            assert_equal(len(blockref.result.hash), 32)
+            assert blockref.result.height > 100
```
In "tests: add functional tests for IPC interface" 592e253f76764f108e9c34fa0da32451091b946a

It will be nice if we use defined constants or use define new ones for the magic numbers, or use response from rpc
It will help in improving the readability of the test.
```diff
diff --git a/test/functional/interface_ipc.py b/test/functional/interface_ipc.py
index 6ff11cbe0f8..912fa3a459b 100755
--- a/test/functional/interface_ipc.py
+++ b/test/functional/interface_ipc.py
@@ -97,18 +97,21 @@ class IPCInterfaceTest(BitcoinTestFramework):
             assert (await mining.result.isInitialBlockDownload(ctx))
             blockref = await mining.result.getTip(ctx)
             assert blockref.hasResult
-            assert_equal(len(blockref.result.hash), 32)
-            assert blockref.result.height > 100
+            block_hash_size = 32
+            assert_equal(len(blockref.result.hash), block_hash_size)
+            current_block_height = self.nodes[0].getchaintips()[0]["height"]
+            assert blockref.result.height == current_block_height
             self.log.debug("Mine a block")
-            wait = mining.result.waitTipChanged(ctx, blockref.result.hash, 1000.0)
+            timeout = 1000.0 # 1000 milliseconds
+            wait = mining.result.waitTipChanged(ctx, blockref.result.hash, )
             self.generate(self.nodes[0], 1)
             newblockref = await wait
-            assert_equal(len(newblockref.result.hash), 32)
-            assert_equal(newblockref.result.height, blockref.result.height + 1)
-            self.log.debug("Wait for timeout (1000 milliseconds)")
-            wait = mining.result.waitTipChanged(ctx, newblockref.result.hash, 1000.0)
+            assert_equal(len(newblockref.result.hash), block_hash_size)
+            assert_equal(newblockref.result.height, current_block_height + 1)
+            self.log.debug("Wait for timeout")
+            wait = mining.result.waitTipChanged(ctx, newblockref.result.hash, timeout)
             oldblockref = await wait
-            assert_equal(len(newblockref.result.hash), 32)
+            assert_equal(len(newblockref.result.hash), block_hash_size)
             assert_equal(oldblockref.result.hash, newblockref.result.hash)
             assert_equal(oldblockref.result.height, newblockref.result.height)
 
@@ -120,7 +123,8 @@ class IPCInterfaceTest(BitcoinTestFramework):
             template = mining.result.createNewBlock(opts)
             self.log.debug("Test some inspectors of Template")
             header = await template.result.getBlockHeader(ctx)
-            assert_equal(len(header.result), 80)
+            block_header_size = 80
+            assert_equal(len(header.result), block_header_size)
             block = await self.parse_and_deserialize_block(template, ctx)
             assert_equal(ser_uint256(block.hashPrevBlock), newblockref.result.hash)
             assert len(block.vtx) >= 1
@@ -134,7 +138,7 @@ class IPCInterfaceTest(BitcoinTestFramework):
             assert_equal(coinbase.vin[0].prevout.hash, 0)
             self.log.debug("Wait for a new template")
             waitoptions = self.capnp_modules['mining'].BlockWaitOptions()
-            waitoptions.timeout = 1000.0
+            waitoptions.timeout = timeout
             waitnext = template.result.waitNext(ctx, waitoptions)
             self.generate(self.nodes[0], 1)
             template2 = await waitnext
```

ismaelsadeeq reviewed (commented):
Code review ACK 235016f5b78ba9f472b56df0825690307fffc7e6 🤖 

Comments are suggestion for improvement and can come in a followup

ryanofsky reviewed (approved):
Code review ACK 235016f5b78ba9f472b56df0825690307fffc7e6. Nice to have this ability to test the interfaces directly. Left a few comments but none are important

ryanofsky reviewed (approved):
Code review ACK db52550507045402e89c6455bec680fcd61e26b6, just making small suggested changes since last review: fixing up pycapnp install instructions, using --break-system-packages only where needed, making small test cleanups

Sjors commented:
re-ACK db52550507045402e89c6455bec680fcd61e26b6

I checked again that the test isn't skipped on the machines where we enable it. It's skipped in the [test each commit job](https://github.com/bitcoin/bitcoin/pull/33201#logs):

```
243/273 - interface_ipc.py skipped (capnp module not available.)
```

That's for the last commit it runs, 1b69f79d138eaf099b108aea6476a0c790d49de7 `ci: enable IPC tests in CI`.

I would be fine with leaving that for a followup.

TheCharlatan commented on test/functional/interface_ipc.py:
```diff
@@ -0,0 +1,174 @@
+            assert_equal(len(block2.vtx), 1)
+            template3 = None
+            last_template = template2
+            if self.uses_wallet:
```
Right, so we might want to call that here?

TheCharlatan replied:
That seems fine, though it would be good to get the fee bumping functionality tested here (in a follow up).

Sjors commented on test/functional/interface_ipc.py:
```diff
@@ -0,0 +1,166 @@
+            self.log.debug("Destroy template objects")
+            template.result.destroy(ctx)
+            template2.result.destroy(ctx)
+            if self.uses_wallet:
```
There was an earlier discussion about that. Apparently in c++ this isn't a problem, but for other language binding we should have defaults. So I'm inclined to agree that we should pick some.

ryanofsky reviewed (approved):
Code review ACK a341e11ac92b30aac856049c696a9ac39854569d. Changes since last review: rebasing, switching to miniwallet and expanding wallet test, improving pycapnp install steps in instructions and CI.

Looking at CI changes here, it's interesting that low-level way we configure CI jobs allows dependencies to be installed that might not be used, and things to be built that are not tested. I think a more ideal config / script boundary might make the config only responsible for specifying what should be tested, and let the script be responsible for figuring out what dependencies need to be installed. That's a wider issue though and CI changes here all look right from what I can tell.

re: libmultiprocess assert failure https://github.com/bitcoin/bitcoin/pull/33201#issuecomment-3258111873, I opened https://github.com/bitcoin-core/libmultiprocess/issues/206 to improve this

sedited reviewed (approved):
ACK a341e11ac92b30aac856049c696a9ac39854569d
